sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs

```diff
@@ -684,6 +685,353 @@ public virtual async Task CloseAsync(CancellationToken cancellationToken = defau
+
+                    foreach (var eventData in eventSet)
+                    {
+                        lastSequence = NextSequence(lastSequence);
+                        eventData.PendingPublishSequenceNumber = lastSequence;
+                    }
+
+                    // Publish the events.
+
+                    cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                    EventHubsEventSource.Log.IdempotentSequencePublish(EventHubName, options.PartitionId, firstSequence, lastSequence);
+                    await SendInternalAsync(eventBatch, cancellationToken).ConfigureAwait(false);
+
+                    // Update state and commit the sequencing.  This needs only to happen at the batch level, as the contained
+                    // events are not accessible by callers.
+
+                    EventHubsEventSource.Log.IdempotentSequenceUpdate(EventHubName, options.PartitionId, partitionState.LastPublishedSequenceNumber.Value, lastSequence);
+                    partitionState.LastPublishedSequenceNumber = lastSequence;
+                    eventBatch.StartingPublishedSequenceNumber = firstSequence;
+                }
+                catch
+                {
+                    // Clear the pending sequence numbers in the face of an exception.
+
+                    foreach (var eventData in eventSet)
+                    {
+                        eventData.PendingPublishSequenceNumber = null;
+                    }
+
+                    throw;
+                }
+                finally
+                {
+                    partitionState.PublishingGuard.Release();
+                    EventHubsEventSource.Log.IdempotentSynchronizationRelease(EventHubName, options.PartitionId);
+                }
+            }
+            catch (Exception ex)
+            {
+                EventHubsEventSource.Log.IdempotentPublishError(EventHubName, options.PartitionId, ex.Message);
+                throw;
+            }
+            finally
+            {
+
+                EventHubsEventSource.Log.IdempotentPublishComplete(EventHubName, options.PartitionId);
+            }
+        }
+
+        /// <summary>
+        ///   Initializes state instance for a given partition.
+        /// </summary>
+        ///
+        /// <param name="partitionState">The state of the partition to be initialized.  This parameter will be mutated by this call.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken" /> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <remarks>
+        ///     The <paramref name="partitionState"/> parameter will be mutated by this call.  To avoid duplicate initialization or state corruption, it
+        ///     is highly recommended that this method only be called while the <see cref="PartitionPublishingState.PublishingGuard" /> primitive of the state
+        ///     instance is held.
```

Comment:
Is there a valid reason to call this without the PartitionPublishingState.PublishingGuard being held? If not, should we ensure that it is?

Reply:
After our offline discussion, I gave this some thought but haven't been able to come up with an approach that I like for validating the semaphore is held by the calling thread.   In the interim, I've strengthened the wording in the comment to remove ambiguity on the expected call pattern and will accept the risk of misuse, since this is private method with very limited scope.